thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/HolidayEventsLoader.java

```diff
@@ -0,0 +1,142 @@
+import com.linkedin.thirdeye.datasource.DAORegistry;
+import java.io.FileInputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The type Holiday events loader, which loads the holiday events from Google Calendar periodically.
+ */
+public class HolidayEventsLoader implements Runnable {
+
+  /**
+   * Instantiates a new Holiday events loader.
+   *
+   * @param configuration the configuration
+   */
+  public HolidayEventsLoader(ThirdEyeAnomalyConfiguration configuration) {
+    calendarList = configuration.getCalendars();
+    keyPath = configuration.getRootDir() + configuration.getKeyPath();
+    scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  /** List of google holiday calendar ids */
+  private static List<String> calendarList;
+  private static String keyPath;
+  private ScheduledExecutorService scheduledExecutorService;
+  private TimeGranularity runFrequency = new TimeGranularity(7, TimeUnit.DAYS);
+
+  private static final Logger LOG = LoggerFactory.getLogger(HolidayEventsLoader.class);
+
+  /** Global instance of the HTTP transport. */
+  private static HttpTransport HTTP_TRANSPORT;
+
+  /** Global instance of the JSON factory. */
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  /** Global instance of the scopes. */
+  private static final Set<String> SCOPES = Collections.singleton(CalendarScopes.CALENDAR_READONLY);
+
+  static {
+    try {
+      HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
+    } catch (Exception e) {
+      LOG.error("Can't create http transport with google api.");
+    }
+  }
+
+  private static final DAORegistry DAO_REGISTRY = DAORegistry.getInstance();
+
+  /**
+   * Start.
+   */
+  public void start() {
+    scheduledExecutorService.scheduleAtFixedRate(this, 0, runFrequency.getSize(), runFrequency.getUnit());
+  }
+
+  /**
+   * Shutdown.
+   */
+  public void shutdown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  /**
+   * Fetch holidays and save to ThirdEye database.
+   */
+  public void run() {
+    EventManager eventDAO = DAO_REGISTRY.getEventDAO();
+
+    List<Event> holidays = getAllHolidays();
+    Set<EventDTO> newEvents = new HashSet<>();
+
+    for (Event holiday : holidays) {
+      EventDTO eventDTO = new EventDTO();
+
+      eventDTO.setStartTime(holiday.getStart().getDate().getValue());
+      eventDTO.setEndTime(holiday.getEnd().getDate().getValue());
+      eventDTO.setName(holiday.getSummary());
+      eventDTO.setEventType(EventType.HOLIDAY.toString());
+
+      newEvents.add(eventDTO);
+    }
+    Set<EventDTO> existingEvents = new HashSet<>(eventDAO.findAll());
```

Comment:
This method is going to become very slow after we collect a lot of holiday events. Please use the notion of watermark. There are two advantages: 1. we could fetch new holiday events that happens after the watermark. 2. watermark of each calendar is a timestamp instead of a list of events.

Reply:
I could do that. But the problem is that there are a lot of existing holiday events in our DB right now. If I don't collect existing events there might be duplication.

Reply:
There's various solutions to this, but I wonder if this would be premature optimization. the database has about 600 events per year. hardly a concern anytime soon